Add support for dynamic DNS updates

tomassedovic opened the pull request:
Passing the update key to `dns_update_key` will cause each node to
register with the specified DNS servers.

tomassedovic committed:
Add support for dynamic DNS updates

Passing the update key to `dns_update_key` will cause each node
(including the dedicated loadbalancer if used) to register with the
specified DNS servers.

jeremyeder commented:
@sjug

tomassedovic committed:
Fix the dns update with multiple servers

The quotes around `%DNS_SERVERS%` meant that the for loop had only a
single item instead of splitting the servers based on the spaces.

markllama commented on fragments/add_dns_record.sh:
```diff
@@ -0,0 +1,26 @@
+
+for DNS_SERVER in "%DNS_SERVERS%"; do
+    # NOTE: the dot after the hostname is necessary
+    /usr/local/bin/update_dns  -s "$DNS_SERVER" -k "$DNS_UPDATE_KEY" "$HOSTNAME." "%IP_ADDRESS%"
```
Because of the way DNS updates work, you must provide the zone even if the name you are setting is a FQDN.

/usr/local/bin/update_dns  -s "$DNS_SERVER" -k "$DNS_UPDATE_KEY" -z "%ZONE%" "$HOSTNAME." "%IP_ADDRESS%"

markllama reviewed (commented):
Update requires the zone even if an FQDN is provided.

markllama commented on fragments/add_dns_record.sh:
```diff
@@ -0,0 +1,26 @@
+
+HOSTNAME="$(hostname --fqdn)"
+
+for DNS_SERVER in %DNS_SERVERS%; do
```
This should really only be a single nameserver.  All of the nameservers should agree and only one needs to be updated.  The others are slaves and will reflect the updates as they come.  See below.

markllama commented on fragments/add_dns_record.sh:
```diff
@@ -0,0 +1,26 @@
+
+for DNS_SERVER in %DNS_SERVERS%; do
+    # NOTE: the dot after the hostname is necessary
+    /usr/local/bin/update_dns  -s "$DNS_SERVER" -k "$DNS_UPDATE_KEY" "$HOSTNAME." "%IP_ADDRESS%"
```
This line also needs the zone (in the heat templates, the `domain_name` value.  with the -z argument.  You must specify what zone to add a name to even if the name is an FQDN because hostnames can have dots in them.

markllama commented on infra.yaml:
```diff
@@ -393,6 +405,17 @@ resources:
+      config:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
```
rather than list_join,  get the first value: `{get_param: [dns_servers, 0]}`

markllama commented on infra.yaml:
```diff
@@ -393,6 +405,17 @@ resources:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
+            '%DNS_UPDATE_KEY%': {get_param: dns_update_key}
+            '%IP_ADDRESS%': {get_attr: [port, fixed_ips, 0, ip_address]}
```
Add '%ZONE%': {get_param: domain_name}

markllama commented on loadbalancer_dedicated.yaml:
```diff
@@ -318,13 +333,25 @@ resources:
+      config:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
```
pass a single value

markllama commented on loadbalancer_dedicated.yaml:
```diff
@@ -318,13 +333,25 @@ resources:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
+            '%DNS_UPDATE_KEY%': {get_param: dns_update_key}
```
Add the domain_name (%ZONE%)

markllama commented on master.yaml:
```diff
@@ -385,6 +397,17 @@ resources:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
+            '%DNS_UPDATE_KEY%': {get_param: dns_update_key}
+            '%IP_ADDRESS%': {get_attr: [port, fixed_ips, 0, ip_address]}
```
add domain_name for the %ZONE%

markllama commented on master.yaml:
```diff
@@ -385,6 +397,17 @@ resources:
+      config:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
```
pass only a single value

markllama commented on node.yaml:
```diff
@@ -500,6 +508,18 @@ resources:
+      config:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
```
Pass a single value {get_param: [dns_servers, 0]}

markllama commented on node.yaml:
```diff
@@ -500,6 +508,18 @@ resources:
+        str_replace:
+          params:
+            '%DNS_SERVERS%': {list_join: [" ", {get_param: dns_servers}]}
+            '%DNS_UPDATE_KEY%': {get_param: dns_update_key}
```
Add domain_name as %ZONE%

markllama reviewed (changes requested):
Only update the first server, the others should be slaves
And provide the zone as well for name updates.

tomassedovic committed:
Specify the zone & only use the first DNS server

The hostnames can have dots in them so it's important to clarify which
zone they belong to.

markllama reviewed (approved):
LGTM
